client/state/add-reducer.ts

```diff
@@ -25,10 +26,10 @@ async function initializeState(
 
 // For a given store, creates a function that adds a new reducer to the store,
 // and loads (asynchronously) and applies the persisted state for it.
-export const addReducerToStore = ( store: object ) => (
-	key: string,
-	reducer: ( state: object, action: object ) => object
-): Promise => {
+export const addReducerToStore = < T extends Reducer >( store: object ) => (
```

Comment:
concerning @sirreal's comment about not finding `dispatch` we might as well use Redux's `Store` type too, no?

```ts
…
import { Reducer, Store } from 'redux';
…

export const addReducerToStore = <T extends Reducer>( store: Store ) => {
	…
}
```

Reply:
I tried that and it pushes the type errors elsewhere…

`Store` doesn't match `object`, `Store` doesn't have `addReducer` method…

What we have is some kind of Enhanced store, that gets more methods added in another file. When I started digging there… well I just got lost.

That demonstrates something I struggle with with gradual typing… how far should we go? Should we expect to have 0 type at any given time? Are there good strategies for addressing problems like these?

Reply:
Yes, that's a good point. It won't take care of the `addReducer` method, but it will take care of `dispatch`. Done.

Reply:
`addReducer` is added by a Redux store enhancer. That's a function that takes an original store and returns an enhanced one. In this case, enhanced with a new method. Therefore, Calypso store's type is an extension of the basic Redux store type.

Redux middlewares are similar, but they enhance only the `dispatch` function. Basic `dispatch` accepts objects with `type` field as actions. Middlewares can change that. `redux-thunk` returns a dispatch function that accepts functions (thunks) as actions, too. Other middlewares can add support for generators.

Store with a middleware has no longer the basic Redux type. `dispatch( action: BaseReduxAction )` gets changed to `dispatch( action: CustomAction )`. Most middlewares only "extend" the `action` type if they modify it at all. But in principle they can change it to something that's unrelated to `BaseReduxAction`.

Reply:
That's better, down to 2 errors for this file 🙂 

```
client/state/add-reducer.ts:33:37 - error TS2339: Property 'storageKey' does not exist on type 'T'.

33  const storageKey: string = reducer.storageKey;
                                       ~~~~~~~~~~

client/state/add-reducer.ts:46:9 - error TS2339: Property 'addReducer' does not exist on type 'Store<any, AnyAction>'.

46  	store.addReducer( key, reducer );
```

Reply:
> That demonstrates something I struggle with with gradual typing… how far should we go? 

Thankfully we only need to go only as far as is helpful at the time. Applications like right here I think are a nice treat on top of other uses because almost no code will be calling this. I'm more concerned about the set of core functions used all throughout Calypso, like the translation hook and `connect()` and typing our React props and selectors.

> Store doesn't match `object`, Store doesn't have `addReducer` method

I'm not sure where `object` comes into play - is that in a test?

This particular problem I think is the inherent frustration with types _that isn't wrong_. That is, a normal fight we have where we say "No, TypeScript, _you_ are wrong!" and then we come back later and realize it was right. Again, we're _depending_ on this property existing when we run this function - why not enforce that constraint so that we can be made aware of it in whatever code we call it in.

---

For the typing we can be graceful if we want. I think ideally we'd type that enhancer that @jsnajdr mentioned because then we could import that type or interface and never worry about getting out of sync with the actual enhancer code.

In the meantime though if we don't want to trace that rabbit hole we can "fake" it by making a basic constraint that we expect such properties and functions/methods on the store.

```ts
export const addReducerToStore = <
	T extends Reducer,
	S extends Store & { addReducer: ( key: string, reducer: T ) => void }
>( store: S ) => {
	…
}
```

We could take this out of the function declaration too if we though it were too noisy, though I find it convenient being in-place right there.

What we've done here is said that whatever is passed in must be _at-least_ a `Store` _and_ it must additionally have this `addReducer` method because _we're going to call it_.

Recently I had a friend ask about a similar situation because he wanted to test the code with a partial mock object. On one hand we could weaken this all the way down to…

```ts
export const addReducerToStore = <T extends Reducer>( store: { addReducer: ( key: string, reducer: T ) => void } ) => { … }
```

…but then we lose some communication value here because we _want_ people to know that this expects a Redux `Store`. In this particular case I recommend what I proposed above but there's  still a sly way of relaxing the typing constraints while still communicating that we expect a store. It's a bit nasty I admit, but _it's serving a specific purpose_.

```ts
type ReducerDeclaration = <R extends Reducer>(key: string, reducer: R) => void;
type ExtendableStore = (Store & { addReducer: ReducerDeclaration }) 

const add = <
  R extends Reducer,
  S extends ExtendableStore,
>(store: Partial<S> & Pick<S, 'addReducer'>) => (k: string[], r: R) => {
  store.addReducer(k.join('.'), r);
}
```
([on the playground](https://www.typescriptlang.org/play/index.html#src=interface%20Store%20%7B%0D%0A%20%20dispatch%3A%20Function%3B%0D%0A%20%20getState%3A%20Function%3B%0D%0A%20%20subscribe%3A%20Function%3B%0D%0A%20%20replaceReducer%3A%20Function%3B%0D%0A%7D%0D%0A%0D%0Atype%20Reducer%20%3D%20%3CS%3E(state%3A%20S%2C%20action%3A%20any)%20%3D%3E%20S%3B%0D%0A%0D%0Atype%20ReducerDeclaration%20%3D%20%3CR%20extends%20Reducer%3E(key%3A%20string%2C%20reducer%3A%20R)%20%3D%3E%20void%3B%0D%0Atype%20ExtendableStore%20%3D%20(Store%20%26%20%7B%20addReducer%3A%20ReducerDeclaration%20%7D)%20%0D%0A%0D%0Aconst%20add%20%3D%20%3C%0D%0A%20%20R%20extends%20Reducer%2C%0D%0A%20%20S%20extends%20ExtendableStore%2C%0D%0A%3E(store%3A%20Partial%3CS%3E%20%26%20Pick%3CS%2C%20'addReducer'%3E)%20%3D%3E%20(k%3A%20string%5B%5D%2C%20r%3A%20R)%20%3D%3E%20%7B%0D%0A%20%20store.addReducer(k.join('.')%2C%20r)%3B%0D%0A%7D))

What I get from this type definition is something like "I expect an extendible store value here but I'm only going to use the `addReducer` method, so that's all I _really_ need." It may have been best to not even suggest this option because it's not _our_ need here but I guess I wanted to go back to "how far should we go" and say "however far our wants and needs lead us."

Reply:
> Property 'storageKey' does not exist on type 'T'

For now how about we say `T extends Reducer & { storageKey: string }`?

Longer-term I think it makes sense to add our own custom Redux types to a file in `state/types.ts` or maybe in the global `client/types.ts` file.

Reply:
Now that I learned some more TypeScript, this came a bit easier. I created some extra interfaces to `&` with, which can at some point be moved to a types file. This should take care of all the existing typing errors.